sql/migrations/0023-orchestration.sql

```diff
@@ -0,0 +1,118 @@
+	device_role text        not null,
+	hardware_id uuid        not null references hardware_product (id),
+
+	created     timestamptz not null default current_timestamp,
+	updated     timestamptz not null default current_timestamp,
+	deactivated bool        not null default false,
+	locked      bool        not null default false,
+	unique(name),
+	unique(name, version),
+	unique(name, version, device_role, hardware_id)
+);
+
+create table orc_lifecycle_plan (
+	orc_lifecycle_id uuid  not null references orc_lifecycle (id),
+	workflow_id      uuid  not null references workflow (id),
+	workflow_order   int   not null,
+	unique(orc_lifecycle_id, workflow_id),
+	unique(orc_lifecycle_id, workflow_order)
+);
+
+/*****************/
+
+create type e_workflow_status as enum (
+	'ongoing',
+	'stopped',
+	'abort',
+	'resume',
+	'completed'
+);
+
+create table workflow_status (
+	id          uuid              primary key default gen_random_uuid(),
+	workflow_id uuid              not null references workflow(id),
+	device_id   text              not null references device(id),
+	timestamp   timestamptz       not null default current_timestamp,
+	status      e_workflow_status not null default 'ongoing'
+);
+
+/*****************/
+
+create table workflow_step (
+	id                 uuid primary key default gen_random_uuid(),
+	created            timestamptz not null default current_timestamp,
+	updated            timestamptz not null default current_timestamp,
+	deactivated        bool not null default false,
+	workflow_id        uuid not null references workflow(id),
+	name               text not null,
+	step_order         int not null,
+	retry              bool default false,
+	max_retries        int not null default 1,
+	validation_plan_id uuid not null, -- fk to real tables when they exist
+	unique(name, workflow_id)
+);
+
+/*****************/
+
+create type e_workflow_step_state as enum (
+	'started',
+	'processing',
+	'complete'
+);
+
+create type e_workflow_validation_status as enum (
+	'pass',
+	'fail',
+	'error',
+	'processing',
+	'noop'
+);
+
+create table workflow_step_status(
+	id                   uuid primary key default gen_random_uuid(),
+	created              timestamptz not null default current_timestamp,
+	updated              timestamptz not null default current_timestamp,
+	device_id            text not null references device(id),
+	workflow_step_id     uuid not null references workflow_step(id),
+	state                e_workflow_step_state not null default 'processing',
+	retry_count          int not null default 1,
+	validation_status    e_workflow_validation_status not null default 'noop',
+	validation_result_id uuid, -- fk to real tables when they exist
+	force_retry          bool default false,
+	overridden            bool default false,
+	data                 jsonb
+);
+
+/*****************/
+
+create or replace view orc_latest_workflow_status as
+select * from (
+	select *, dense_rank() over (
+		partition by workflow_id
```

Comment:
Same issue still exists, unless this is what you intend. You mention in the POD

> This can be used, for instance, to find all workflows that are ONGOING.

This query returns the most recent workflow status per workflow. The most recent status may not be ONGOING.